src/csharp/Microsoft.Spark/Sql/StorageLevel.cs

```diff
@@ -0,0 +1,91 @@
+    /// data in memory in a JAVA-specific serialized format, and whether to replicate the RDD
+    /// partitions on multiple nodes.Also contains static constants for some commonly used storage
+    /// levels, MEMORY_ONLY. Since the data is always serialized on the Python side, all the
+    /// constants use the serialized formats.
+    /// </summary>
+    public sealed class StorageLevel : IJvmObjectReferenceProvider
+    {
+        private readonly JvmObjectReference _jvmObject;
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        public bool UseDisk { get; private set; }
+        public bool UseMemory { get; private set; }
+        public bool UseOffHeap { get; private set; }
+        public bool Deserialized { get; private set; }
+        public int Replication { get; private set; }
+
+        internal StorageLevel(JvmObjectReference jvmObject)
+        {
+            UseDisk = (bool)jvmObject.Invoke("useDisk");
+            UseMemory = (bool)jvmObject.Invoke("useMemory");
+            UseOffHeap = (bool)jvmObject.Invoke("useOffHeap");
+            Deserialized = (bool)jvmObject.Invoke("deserialized");
+            Replication = (int)jvmObject.Invoke("replication");
+            _jvmObject = jvmObject;
+        }
+
+        public StorageLevel(
+            bool useDisk,
+            bool useMemory,
+            bool useOffHeap,
+            bool deserialized,
+            int replication = 1)
+        {
+            UseDisk = useDisk;
+            UseMemory = useMemory;
+            UseOffHeap = useOffHeap;
+            Deserialized = deserialized;
+            Replication = replication;
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(
+                "org.apache.spark.storage.StorageLevel",
+                UseDisk,
+                UseMemory,
+                UseOffHeap,
+                Deserialized,
+                Replication);
+        }
+
+        public string Description() => (string)_jvmObject.Invoke("description");
+
+        public override string ToString() => (string)_jvmObject.Invoke("toString");
+
+        public static StorageLevel DISK_ONLY { get; } = 
+            new StorageLevel(true, false, false, false);
+        public static StorageLevel DISK_ONLY_2 { get; } = 
+            new StorageLevel(true, false, false, false, 2);
+        public static StorageLevel MEMORY_ONLY { get; } = 
+            new StorageLevel(false, true, false, false);
+        public static StorageLevel MEMORY_ONLY_2 { get; } = 
+            new StorageLevel(false, true, false, false, 2);
+        public static StorageLevel MEMORY_AND_DISK { get; } = 
+            new StorageLevel(true, true, false, false);
+        public static StorageLevel MEMORY_AND_DISK_2 { get; } = 
+            new StorageLevel(true, true, false, false, 2);
+        public static StorageLevel OFF_HEAP { get; } = 
+            new StorageLevel(true, true, true, false, 1);
```

Comment:
group with the other properties.  Also, should we instantiate these lazily ?

Reply:
Done, I think instantiating lazily makes sense. @imback82  thoughts?